feat(router): add skipNavigation option to routerLink

alexduer opened the pull request:
The RouterLink directive now has a 'skipNavigation' boolean which disables navigating.

Closes #31154
Closes #13980

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

Issue Number: 31154


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
I could not find documentation for the 'skipLocationChange' boolean, so I did not add any documentation for the 'skipNavigation' boolean either.

atscott commented on packages/router/src/directives/router_link.ts:
```diff
@@ -164,7 +166,9 @@ export class RouterLink {
       replaceUrl: attrBoolValue(this.replaceUrl),
     };
-    this.router.navigateByUrl(this.urlTree, extras);
+    if (!attrBoolValue(this.skipNavigation)) {
```
I think this would be better as an early return at the top like what's done with the `onClick` of `RouterLinkWithHref`

atscott commented on packages/router/src/directives/router_link.ts:
```diff
@@ -262,7 +268,9 @@ export class RouterLinkWithHref implements OnChanges, OnDestroy {
       state: this.state
     };
-    this.router.navigateByUrl(this.urlTree, extras);
+    if (!attrBoolValue(this.skipNavigation)) {
```
Move this to be an early return at the top with the others. Additionally, I think you should be returning `true` instead of `false` if you're not navigating.

atscott commented on packages/router/src/directives/router_link.ts:
```diff
@@ -124,6 +124,8 @@ export class RouterLink {
   // TODO(issue/24571): remove '!'.
   @Input() skipLocationChange !: boolean;
   // TODO(issue/24571): remove '!'.
+  @Input() skipNavigation !: boolean;
```
I'm not sure how I feel about `skipNavigation` since navigating is kind of the only thing a `routerLink` *does* in the first place. Maybe `disabled` or `disableNavigation` would be a more appropriate name. What do you think?

alexduer replied:
Good point! 'disableNavigation' is definitely clearer.

alexduer committed:
feat(router): add disableNavigation option to routerLink

The RouterLink directive now has a 'disableNavigation' boolean which disables navigating.

Closes #31154

atscott commented:
Can you take a look at the CI failures and try to fix them? Also adding @gkalpak for some input on how this should be documented, if merged.

atscott commented on packages/router/src/directives/router_link.ts:
```diff
@@ -124,6 +124,8 @@ export class RouterLink {
   // TODO(issue/24571): remove '!'.
   @Input() skipLocationChange !: boolean;
   // TODO(issue/24571): remove '!'.
+  @Input() disableNavigation !: boolean;
```
Rather than continuing to use `!` and adding to the technical debt for #24571, how about using `?` since Inputs are optional?

jdgarvey commented:
@alexduer are you still working on this? If you are that's cool, but if you aren't I can pick this up and get it through the rest of the process.

alexduer commented:
@jdgarvey you are welcome to pick up the issue

atscott commented:
Closing, as it looks like the work was handed off and is now in #39916
